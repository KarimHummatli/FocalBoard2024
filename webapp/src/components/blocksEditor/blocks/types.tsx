import {ContentBlockTypes} from '../../../blocks/block'

export type BlockInputProps = {
    onChange: (value: string) => void
    value: string
    onCancel: () => void
    onSave: (val: string) => void
}

export type ContentType = {
    name: ContentBlockTypes
    displayName: string
    slashCommand: string
    prefix: string
    Input: React.FunctionComponent<BlockInputProps>
    render: (value: string) => React.ReactNode
    runSlashCommand: (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void, ...args: string[]) => void
    nextType?: string
}
