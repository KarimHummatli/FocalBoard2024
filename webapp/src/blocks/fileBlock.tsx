// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Block, createBlock} from './block'
import {ContentBlock} from './contentBlock'

type FileBlockFields = {
    attachmentId: string
    attachmentName: string
    attachmentType: string
    attachmentSize: number
    isAttachment: boolean
}

type FileBlock = ContentBlock & {
    type: 'attachment'
    fields: FileBlockFields
}

function createFileBlock(block?: Block): FileBlock {
    return {
        ...createBlock(block),
        type: 'attachment',
        fields: {
            attachmentId: block?.fields.attachmentId || '',
            attachmentName: block?.fields.attachmentName || '',
            attachmentType: block?.fields.attachmentType || '',
            attachmentSize: block?.fields.attachmentSize || '',
            isAttachment: block?.fields.isAttachment || false,
        },
    }
}

export {FileBlock, createFileBlock}
