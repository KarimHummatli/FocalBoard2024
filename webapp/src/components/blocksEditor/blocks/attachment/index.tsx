import React, {useRef, useEffect} from 'react'
import {BlockInputProps, ContentType} from '../types'

import './attachment.scss'

const Attachment: ContentType = {
    name: 'attachment',
    displayName: 'Attachment',
    slashCommand: '/attachment',
    prefix: '',
    runSlashCommand: (): void => {},
    editable: false,
    Display: (props: BlockInputProps) => <div className='AttachmentView'>ðŸ“Ž {props.value}</div>,
    Input: (props: BlockInputProps) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.click()
        }, [])

        return (
            <input
                ref={ref}
                className='Attachment'
                type='file'
                onChange={(e) => {
                    const files = e.currentTarget?.files
                    if (files) {
                        for (let i = 0; i < files.length; i++) {
                            const file = files.item(i)
                            if (file) {
                                props.onSave(file.name as string)
                            }
                        }
                    }
                }}
            />
        )
    }
}

Attachment.runSlashCommand = (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void): void => {
    changeType(Attachment)
    changeValue('')
}

export default Attachment
