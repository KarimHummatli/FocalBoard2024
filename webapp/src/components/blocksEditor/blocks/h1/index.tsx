import React, {useRef, useEffect} from 'react'
import {marked} from 'marked'

import {BlockInputProps, ContentType} from '../types'

import './h1.scss'

const H1: ContentType = {
    name: 'h1',
    displayName: 'Title',
    slashCommand: '/title',
    prefix: '# ',
    render: (value: string) => {
        const renderer = new marked.Renderer()
        const html = marked('# '+value, {renderer, breaks: true})
        return <div
            dangerouslySetInnerHTML={{__html: html.trim()}}
        />
    },
    runSlashCommand: (): void => {},
    Input: (props: BlockInputProps) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.focus()
        }, [])
        return (
            <input
                ref={ref}
                className='H1'
                onChange={(e) => props.onChange(e.currentTarget.value)}
                onKeyDown={(e) => {
                    if (props.value === '' && e.key === "Backspace") {
                        props.onCancel()
                    }
                    if (e.key === "Enter") {
                        props.onSave(props.value)
                    }
                }}
                value={props.value}
            />
        )
    }
}

H1.runSlashCommand = (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void, ...args: string[]): void => {
    changeType(H1)
    changeValue(args.join(' '))
}

export default H1
