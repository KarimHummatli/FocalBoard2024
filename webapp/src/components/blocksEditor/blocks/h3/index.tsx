import React, {useRef, useEffect} from 'react'
import {marked} from 'marked'

import {BlockInputProps, ContentType} from '../types'

import './h3.scss'

const H3: ContentType = {
    name: 'h3',
    displayName: 'Sub Sub title',
    slashCommand: '/subsubtitle',
    prefix: '### ',
    render: (value: string) => {
        const renderer = new marked.Renderer()
        const html = marked('### '+value, {renderer, breaks: true})
        return <div
            dangerouslySetInnerHTML={{__html: html.trim()}}
        />
    },
    runSlashCommand: (): void => {},
    Input: (props: BlockInputProps) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.focus()
        }, [])
        return (
            <input
                ref={ref}
                className='H3'
                onChange={(e) => props.onChange(e.currentTarget.value)}
                onKeyDown={(e) => {
                    if (props.value === '' && e.key === "Backspace") {
                        props.onCancel()
                    }
                    if (e.key === "Enter") {
                        props.onSave(props.value)
                    }
                }}
                value={props.value}
            />
        )
    }
}

H3.runSlashCommand = (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void, ...args: string[]): void => {
    changeType(H3)
    changeValue(args.join(' '))
}

export default H3
