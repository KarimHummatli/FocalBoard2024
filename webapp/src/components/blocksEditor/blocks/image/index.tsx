import React, {useRef, useEffect} from 'react'
import {BlockInputProps, ContentType} from '../types'

import './image.scss'

type FileInfo = {
    file: string|ArrayBuffer
    width?: number
    align?: 'left'|'center'|'right'
}

const Image: ContentType<FileInfo> = {
    name: 'image',
    displayName: 'Image',
    slashCommand: '/image',
    prefix: '',
    runSlashCommand: (): void => {},
    editable: false,
    Display: (props: BlockInputProps<FileInfo>) => {
        if (props.value.file) {
            return <img className='ImageView' src={props.value.file as string}/>
        }
        return null
    },
    Input: (props: BlockInputProps<FileInfo>) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.click()
        }, [])

        return (
            <div>
                {props.value.file && (typeof props.value.file === 'string') && (
                    <img
                        className='ImageView' src={props.value.file}
                        onClick={() => ref.current?.click()}
                    />
                )}
                <input
                    ref={ref}
                    className='Image'
                    type='file'
                    accept='image/*'
                    onChange={(e) => {
                        const reader = new FileReader();
                        const file = (e.currentTarget?.files || [])[0]
                        reader.readAsArrayBuffer(file);
                        reader.onload = function () {
                            props.onSave({file: reader.result as ArrayBuffer})
                        }
                    }}
                />
            </div>
        )
    }
}

Image.runSlashCommand = (changeType: (contentType: ContentType<FileInfo>) => void, changeValue: (value: FileInfo) => void): void => {
    changeType(Image)
    changeValue({file: ''})
}

export default Image
