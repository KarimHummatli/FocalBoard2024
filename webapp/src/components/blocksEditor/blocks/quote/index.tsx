import React, {useRef, useEffect} from 'react'
import {marked} from 'marked'

import {BlockInputProps, ContentType} from '../types'

import './quote.scss'

const Quote: ContentType = {
    name: 'quote',
    displayName: 'Quote',
    slashCommand: '/quote',
    prefix: '> ',
    render: (value: string) => {
        const renderer = new marked.Renderer()
        const html = marked('> '+value, {renderer, breaks: true})
        return <div
            className='Quote'
            dangerouslySetInnerHTML={{__html: html.trim()}}
        />
    },
    runSlashCommand: (): void => {},
    Input: (props: BlockInputProps) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.focus()
        }, [])
        return (
            <blockquote className='Quote'>
                <input
                    ref={ref}
                    onChange={(e) => props.onChange(e.currentTarget.value)}
                    onKeyDown={(e) => {
                        if (props.value === '' && e.key === "Backspace") {
                            props.onCancel()
                        }
                        if (e.key === "Enter") {
                            props.onSave(props.value)
                        }
                    }}
                    onBlur={() => props.onSave(props.value)}
                    value={props.value}
                />
            </blockquote>
        )
    }
}

Quote.runSlashCommand = (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void, ...args: string[]): void => {
    changeType(Quote)
    changeValue(args.join(' '))
}

export default Quote
