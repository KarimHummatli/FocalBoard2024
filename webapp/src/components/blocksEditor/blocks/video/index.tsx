import React, {useRef, useEffect} from 'react'
import {BlockInputProps, ContentType} from '../types'

import './video.scss'

const Video: ContentType = {
    name: 'video',
    displayName: 'Video',
    slashCommand: '/video',
    prefix: '',
    render: (value: string) => (
        <video width="320" height="240" controls className='VideoView'>
            <source src={value} type='video/mp4'/>
        </video>
    ),
    runSlashCommand: (): void => {},
    Input: (props: BlockInputProps) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.click()
        }, [])

        return (
            <input
                ref={ref}
                className='Video'
                type='file'
                accept='video/*'
                onChange={(e) => {
                    const reader = new FileReader();
                    const file = (e.currentTarget?.files || [])[0]
                    reader.readAsDataURL(file);
                    reader.onload = function () {
                        props.onSave(reader.result as string)
                    }
                }}
            />
        )
    }
}

Video.runSlashCommand = (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void): void => {
    changeType(Video)
    changeValue('')
}

export default Video
