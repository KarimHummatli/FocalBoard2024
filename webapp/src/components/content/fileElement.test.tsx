// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react'

import {render} from '@testing-library/react'

import {act} from 'react-dom/test-utils'

import {mocked} from 'jest-mock'

import {FileBlock} from '../../blocks/fileBlock'

import {wrapIntl} from '../../testUtils'

import octoClient from '../../octoClient'

import FileElement from './fileElement'

jest.mock('../../octoClient')
const mockedOcto = mocked(octoClient, true)
mockedOcto.getFileAsDataUrl.mockResolvedValue({url: 'test.txt'})
mockedOcto.getFileInfo.mockResolvedValue({
    name: 'test.txt',
    size: 2300,
    extension: '.txt',
})

describe('component/content/FileBlock', () => {
    const defaultBlock: FileBlock = {
        id: 'test-id',
        boardId: '1',
        parentId: '',
        modifiedBy: 'test-user-id',
        schema: 0,
        type: 'attachment',
        title: 'test-title',
        fields: {
            attachmentId: 'test.txt',
        },
        createdBy: 'test-user-id',
        createAt: 0,
        updateAt: 0,
        deleteAt: 0,
        limited: false,
    }

    test('should match snapshot', async () => {
        const component = wrapIntl(
            <FileElement
                block={defaultBlock}
            />,
        )
        let fileContainer: Element | undefined
        await act(async () => {
            const {container} = render(component)
            fileContainer = container
        })
        expect(fileContainer).toMatchSnapshot()
    })

    test('archived file', async () => {
        mockedOcto.getFileAsDataUrl.mockResolvedValue({
            archived: true,
            name: 'FileName',
            extension: '.txt',
            size: 165002,
        })

        const component = wrapIntl(
            <FileElement
                block={defaultBlock}
            />,
        )
        let fileContainer: Element | undefined
        await act(async () => {
            const {container} = render(component)
            fileContainer = container
        })
        expect(fileContainer).toMatchSnapshot()
    })
})
