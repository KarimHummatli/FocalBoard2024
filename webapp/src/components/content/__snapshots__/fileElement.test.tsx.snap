// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/content/FileBlock archived file 1`] = `<div />`;

exports[`component/content/FileBlock should match snapshot 1`] = `
<div>
  <div
    class="FileElement mr-4"
  >
    <div
      class="fileElement-icon-division"
    >
      <i
        class="CompassIcon icon-file-text-outline-large fileElement-icon"
      />
    </div>
    <div
      class="fileElement-file-details mt-3"
    >
      <div
        class="fileElement-file-name"
      >
        test.txt
      </div>
      <div
        class="fileElement-file-ext-and-size"
      >
        txt
         
        2.2 KiB
      </div>
    </div>
    <div
      class="fileElement-delete-download"
    >
      <div
        aria-label="menuwrapper"
        class="MenuWrapper mt-3 fileElement-menu-icon"
        role="button"
      >
        <button
          class="IconButton size--medium"
          type="button"
        >
          <i
            class="CompassIcon icon-dots-vertical"
          />
        </button>
      </div>
      <a
        class="fileElement-download-btn mt-5 mr-2"
        download="test.txt"
        href="test.txt"
        rel="noopener noreferrer"
        target="_blank"
      >
        <i
          class="CompassIcon icon-download-outline"
        />
      </a>
    </div>
  </div>
</div>
`;
